src/analysis/smoment.jl

```diff
@@ -1,75 +1,84 @@
-    )
 
-Construct an `SMomentModel` model using `model` and `rid_isozymes`.
 """
-function make_smomentmodel(
-    model::StandardModel;
-    rid_isozymes = Dict{String,Vector{Isozyme}}(),
-    enzyme_capacity = 0.0,
-)
-
-    # check that input data is in correct format for smoment
-    if any(length(v) > 1 for v in values(rid_isozymes))
-        @warn(
-            "For SMOMENT to work correctly, no isozymes are allowed. Call `remove_slow_isozymes!` to fix the input data."
-        )
-    end
+    make_smoment_model(
+        model::MetabolicModel;
+        reaction_isozymes::Function,
+        gene_product_molar_mass::Function,
+        total_enzyme_capacity::Float64,
+    )
 
-    irrevS, lb_fluxes, ub_fluxes, reaction_map, metabolite_map =
-        _build_irreversible_stoichiometric_matrix(model)
+Construct a model with a structure given by sMOMENT algorithm; returns a
+[`SMomentModel`](@ref) (see the documentation for details.
 
-    #: size of resultant model
-    num_reactions = size(irrevS, 2)
-    num_metabolites = size(irrevS, 1)
-    num_vars = num_reactions + 1
+`reaction_isozymes` parameter is a function that returns a single isozyme for
+each reaction, or `nothing` if the reaction is not enzymatic. If the reaction
+has multiple isozymes, use [`smoment_isozyme_score`](@ref) to select the "best"
+one, as recommended by the sMOMENT approach.
 
-    #: equality lhs
-    Se = zeros(1, num_reactions)
+`gene_product_molar_mass` parameter is a function that returns a molar mass of
+each gene product (relative to `total_enzyme_capacity` and the specified
+kcats), as specified by sMOMENT.
 
-    for (rid, col_idx) in reaction_map
-        original_rid = string(split(rid, "§")[1])
+`total_enzyme_capacity` is the maximum "enzyme capacity" consumption of the
+model.
+"""
+function make_smoment_model(
+    model::MetabolicModel;
+    reaction_isozymes::Function,
+    gene_product_molar_mass::Function,
+    total_enzyme_capacity::Float64,
+)
+    columns = Vector{_smoment_column}()
+    coupling_row_reaction = Int[]
 
-        # skip these entries
-        !haskey(rid_isozymes, original_rid) && continue
-        # these entries have kcats, only one GRR by assumption
-        isozyme = first(rid_isozymes[original_rid])
-        mw = sum([model.genes[gid].molar_mass * ps for (gid, ps) in isozyme.stoichiometry])
-        kcat = contains(rid, "§FOR") ? first(isozyme.kcats) : last(isozyme.kcats)
-        Se[1, col_idx] = -mw / kcat
-    end
+    (lbs, ubs) = bounds(model)
+    rids = reactions(model)
 
-    S = [
-        irrevS zeros(num_metabolites, 1)
-        Se 1.0
-    ]
+    for i = 1:n_reactions(model)
+        isozyme = reaction_isozymes(rids[i])
+        if isnothing(isozyme)
+            # non-enzymatic reaction (or a totally ignored one)
+            push!(columns, _smoment_column(i, 0, 0, lbs[i], ubs[i], 0))
+            continue
+        end
+        # pick a new row for "arm reaction" coupling
+        push!(coupling_row_reaction, i)
```

Comment:
the "arm" here is the thing that binds the 2 sides together to provide the "overall" bound. Thinking about it, it's likely redundant here, because 2 reactions can't realistically overshoot the overall maxima and minima, right?

Reply:
yeah, I think so